component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/service/LocalCacheService.java

```diff
@@ -17,71 +17,236 @@
 import org.talend.sdk.component.api.service.cache.LocalCache;
+import org.talend.sdk.component.api.service.configuration.Configuration;
 import org.talend.sdk.component.runtime.serialization.SerializableService;
 
-import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
 
-@AllArgsConstructor
-public class LocalCacheService implements LocalCache, Serializable {
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toList;
 
+
+/**
+ * Implementation of LocalCache with in memory conccurent map.
+ * @param <T>   the type of cached data.
+ */
+@RequiredArgsConstructor
+public class LocalCacheService<T> implements LocalCache<T>, Serializable {
+
+    /** plugin name for this cache */
     private final String plugin;
 
-    private final ConcurrentMap<String, Element> cache = new ConcurrentHashMap<>();
 
+    private final ConcurrentMap<String, ElementImpl<T>> cache = new ConcurrentHashMap<>();
+
+    @Configuration("talend.component.manager.services.cache.eviction")
+    private Supplier<CacheConfiguration> configuration;
+
+    // scheduler we use to evict tokens
+    private volatile ScheduledExecutorService threadService;
+
+    /**
+     * Evict an object.
+     * @param key key to evict.
+     */
     @Override
     public void evict(final String key) {
-        cache.remove(internalKey(key));
+        final ElementImpl<T> removedElement = cache.remove(internalKey(key));
+        if (removedElement != null) {
+            removedElement.release();
+        }
     }
 
     @Override
-    public <T> void evictIfValue(final String key, final T expected) {
-        cache.remove(internalKey(key), new Element(expected, 0));
+    public void evictIfValue(final String key, final T expected) {
+        final String realKey = internalKey(key);
+
+        // use compute to be able to call release.
+        cache.compute(realKey,
+                (String oldKey , ElementImpl<T> oldElement) -> {
+                    if (oldElement != null && Objects.equals(oldElement.getValue(), expected)) {
+                        // ok to evit, so do release.
+                        oldElement.release();
+                        return null;
+                    }
+                    return oldElement;
+                });
+
     }
 
+
+
     @Override
-    public <T> T computeIfAbsent(final String key, final long timeoutMs, final Supplier<T> value) {
+    public T computeIfAbsent(String key,
+            final Predicate<Element<T>> toRemove,
+            final Supplier<T> value) {
+
+        final ElementImpl<T> element = this.addToMap(key,
+                () -> new ElementImpl<>(value, toRemove, -1, null));
+
+        return element.value;
+    }
+
+    @Override
+    public T computeIfAbsent(String key, long timeoutMs, Supplier<T> value) {
+
+        final long endOfValidity = timeoutMs > 0 ? System.currentTimeMillis() + timeoutMs : -1;
```

Comment:
we should likely move to Clock, this would enable to test it moreeasily
In the default service provider it can be the system utc default one but in test you will pass the "test"/settable one.
Will avoid the "sleeps".

Reply:
All Clock subclasses are immutable, so i can't modify time.
With a mutable Clock class, i can easily play with endOfValidity field.
But how with ScheduledExecutorService.schedule ?